### PR TITLE
Breadcrumbs for database operations

### LINKED ISSUE #1341: Breadcrumbs for database operations
- Meta Issue https://github.com/getsentry/team-mobile/issues/92

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- Create breadcrumbs for database operations where also spans are created.
- Set the following data:
  - type
    - query for database queries
    - error for errors
    - nothing in other cases
  - category
    - same as span operation
  - data
    - status
      - either ok or error
    - database system
    - database name
  - message
    - sqlite queries/commands from internal buffer (e.g: 'UPDATE Product SET title = ?')

- Also bumped the iOS simulator to 17, because the 16.4 is invalid in the most recent macOS-13 runner update.

## :bulb: Motivation and Context

Closes #1341

## :green_heart: How did you test it?

Added unit tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

@buenaflor @krystofwoldrich Please check if the breadcrumb data makes sense.

### CODE REVIEW

File: sqflite/lib/src/sentry_batch.dart [hunk 1]
```diff
@@ -64,14 +70,24 @@ class SentryBatch implements Batch {
...
         span?.status = SpanStatus.ok();
 
+        breadcrumb.data?['status'] = 'ok';
+
         return result;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
 
+        breadcrumb.data?['status'] = 'internalError';
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_batch.dart]: shouldn't it be `internal_error`?

File: sqflite/lib/src/sentry_batch.dart [hunk 2]
```diff
@@ -64,14 +70,24 @@ class SentryBatch implements Batch {
...
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_batch.dart]: type shouldn't be `error` here

File: sqflite/lib/src/sentry_batch.dart [hunk 3]
```diff
@@ -102,15 +124,24 @@ class SentryBatch implements Batch {
...
+        breadcrumb.data?['status'] = 'ok';
 
         return result;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
 
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_batch.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database.dart [hunk 4]
```diff
@@ -75,24 +75,39 @@ class SentryDatabase extends SentryDatabaseExecutor implements Database {
...
         await _database.close();
 
         span?.status = SpanStatus.ok();
+        breadcrumb.data?['status'] = 'ok';
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database.dart [hunk 5]
```diff
@@ -152,15 +175,24 @@ class SentryDatabase extends SentryDatabaseExecutor implements Database {
...
         span?.status = SpanStatus.ok();
+        breadcrumb.data?['status'] = 'ok';
 
         return result;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 6]
```diff
@@ -47,20 +47,35 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
             await _executor.delete(table, where: where, whereArgs: whereArgs);
 
         span?.status = SpanStatus.ok();
+        breadcrumb.data?['status'] = 'ok';
 
         return result;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
+        breadcrumb.data?['status'] = 'internalError';
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 7]
```diff
@@ -78,17 +93,36 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
         await _executor.execute(sql, arguments);
 
         span?.status = SpanStatus.ok();
+        breadcrumb.data?['status'] = 'ok';
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 8]
```diff
@@ -78,17 +93,36 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
+          level: SentryLevel.error,
+        );
+        breadcrumb = breadcrumb.copyWith(
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: guess these are duplicated lines, right?

File: sqflite/lib/src/sentry_database_executor.dart [hunk 9]
```diff
@@ -125,15 +166,23 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
         span?.status = SpanStatus.ok();
+        breadcrumb.data?['status'] = 'ok';
 
         return result;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 10]
```diff
@@ -188,15 +245,23 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 11]
```diff
@@ -253,15 +326,23 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 12]
```diff
@@ -278,19 +359,34 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 13]
```diff
@@ -307,19 +403,34 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 14]
```diff
@@ -339,19 +450,35 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 15]
```diff
@@ -380,15 +515,23 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 16]
```diff
@@ -405,19 +548,34 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_database_executor.dart [hunk 17]
```diff
@@ -457,15 +622,23 @@ class SentryDatabaseExecutor implements DatabaseExecutor {
...
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_database_executor.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_sqflite.dart [hunk 18]
```diff
@@ -40,28 +40,44 @@ Future<Database> openDatabaseWithSentry(
...
       span?.status = SpanStatus.ok();
+      breadcrumb.data?['status'] = 'ok';
+
       return sentryDatabase;
     } catch (exception) {
       span?.throwable = exception;
       span?.status = SpanStatus.internalError();
+      breadcrumb.data?['status'] = 'internalError';
+      breadcrumb = breadcrumb.copyWith(
+        type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_sqflite.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/sentry_sqflite_database_factory.dart [hunk 19]
```diff
@@ -59,30 +59,45 @@ class SentrySqfliteDatabaseFactory with SqfliteDatabaseFactoryMixin {
...
+        breadcrumb.data?['status'] = 'ok';
+
         return sentryDatabase;
       } catch (exception) {
         span?.throwable = exception;
         span?.status = SpanStatus.internalError();
-
+        breadcrumb.data?['status'] = 'internalError';
+        breadcrumb = breadcrumb.copyWith(
+          type: 'error',
```
Comment (stefanosiano) [inline sqflite/lib/src/sentry_sqflite_database_factory.dart]: same comment on type and `internal_error` as above

File: sqflite/lib/src/utils/sentry_database_span_attributes.dart [hunk 20]
```diff
@@ -10,3 +10,12 @@ void setDatabaseAttributeData(ISentrySpan? span, String? dbName) {
     span?.setData(SentryDatabase.dbNameKey, dbName);
   }
 }
+
+/// Sets the database attributes on the [breadcrumb].
+/// It contains the database system and the database name.
+void setDatabaseAttributeOnBreadcrumb(Breadcrumb breadcrumb, String? dbName) {
+  breadcrumb.data?[SentryDatabase.dbSystemKey] = SentryDatabase.dbSystem;
```
Comment (stefanosiano) [inline sqflite/lib/src/utils/sentry_database_span_attributes.dart:17]: [Docs](https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/#breadcrumb-types) suggest to set the breadcrumb type for database operation to `query`, even if don't have that for iOS, Android or RN
Let's set it here, as this is used for all the database breadcrumbs

**Review by stefanosiano — CHANGES_REQUESTED** [review]
Just a comment on the type and the `internal_error` status

File: sqflite/lib/src/sentry_batch.dart (hunk 2 above)
Comment (denrase) [reply sqflite/lib/src/sentry_batch.dart]: I checked in the documentation and error seemed appropriate, or am I missing something here?

https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/#breadcrumb-types

Do you have a suggestion which type to use instead?

**Review by buenaflor — COMMENTED** [review]
lgtm although we first need clarification on Stefano's error comments. To me it seems like `error` looks fine as type when catching the exception

File: sqflite/lib/src/sentry_batch.dart (hunk 2 above)
Comment (stefanosiano) [reply sqflite/lib/src/sentry_batch.dart]: my understanding is the `error` type is for Exceptions.
Looking at what is done for [Java OkHttp](https://github.com/getsentry/sentry-java/blob/12704a922bb5b964a9f2f88753acffaeee85ff14/sentry-android-okhttp/src/main/java/io/sentry/android/okhttp/SentryOkHttpInterceptor.kt#L142), we send a breadcrumb with type `http`, regardless the status code.
Let me ask the team and get back to you

Comment (stefanosiano) [reply sqflite/lib/src/sentry_batch.dart]: Confirmed: the `query` type is used to report a query is being executed. it has nothing to do with its result.
The `error` type is used for exceptions only
Also, we could change the level to `warning`, as the exception in the query does not necessarily means a crash of the app, or does it?

**Comment by stefanosiano** [comment]
> lgtm although we first need clarification on Stefano's error comments. To me it seems like `error` looks fine as type when catching the exception

the breadcrumb is still about the query, not the exception, so the [type shouldn't change](https://github.com/getsentry/sentry-dart/pull/1656#discussion_r1363524921)

File: sqflite/lib/src/sentry_batch.dart (hunk 2 above)
Comment (denrase) [reply sqflite/lib/src/sentry_batch.dart]: Thank you for clearing this up! Warning seems better yes, as users might handle these.

Comment (denrase) [reply sqflite/lib/src/sentry_batch.dart]: Ok updated them to query. Thinking if this is also correct in `apply` and `commit`, which are write operations. 🤔

File: sqflite/lib/src/utils/sentry_database_span_attributes.dart (hunk 20 above)
Comment (denrase) [reply sqflite/lib/src/utils/sentry_database_span_attributes.dart:17]: Need to make a copy since the type of the breadcrumb is immutable.

Comment (denrase) [reply sqflite/lib/src/utils/sentry_database_span_attributes.dart:17]: I've added the type in the ctor

**Comment by denrase** [comment]
@stefanosiano Updated everything as suggested. Just a question if type `query` should be set for write operations like `commit` & `apply`

File: sqflite/lib/src/sentry_batch.dart (hunk 2 above)
Comment (buenaflor) [reply sqflite/lib/src/sentry_batch.dart]: @stefanosiano any thoughts on that? not sure about that here

**Comment by denrase** [comment]
@stefanosiano WDYT, can we merge or is there something left here?

File: sqflite/lib/src/sentry_sqflite.dart (hunk 18 above)
Comment (stefanosiano) [reply sqflite/lib/src/sentry_sqflite.dart]: I'd remove this `type: 'error'` from here

File: sqflite/lib/src/sentry_batch.dart (hunk 2 above)
Comment (stefanosiano) [reply sqflite/lib/src/sentry_batch.dart]: yeah, I think it's fine, as they are still operations on the database

**Comment by stefanosiano** [comment]
@denrase There are just a couple of `type: 'error'` left that i'd remove
after that we can merge

**Comment by denrase** [comment]
@stefanosiano Sry about that, they are removed now.

**Review by stefanosiano — APPROVED** [review]

**Pull request merged**